### PR TITLE
nixos/kubernetes: Patch kube-proxy for iptables 1.6.2 compatibility

### LINKED ISSUE #35544: kube-proxy is broken by iptables 1.6.1 -> 1.6.2
## Issue description
This is mostly a reminder to follow up on what upstream decides to do.

Since iptables 1.6.2, iptables-restore fails instead of ignoring erroneous arguments, specifically `-w5` which kube-proxy has been (erroneously) using instead of `-w 5`. This causes its firewall rules to fail along with k8s networking in general.

https://github.com/kubernetes/kubernetes/issues/58956

### Steps to reproduce
Run eg. the dns.multinode test.

### PR DESCRIPTION
**Pull request by srhb**
###### Motivation for this change
Fixes #35544 

###### Things done
Upstream patch to insert the missing space in the "-w 5" argument to iptables-restore included. This fixes kube-proxy.

This should be backported to 18.03

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
@GrahamcOfBorg build kubernetes

**Comment by srhb** [comment]
@Mic92 Any idea how to make grahamcofborg run the tests as well? I can't seem to figure out the right command, something appears to be different for the kubernetes test set (and it's not running on Hydra either, as far as I can see.)

**Comment by srhb** [comment]
Hm, I'll look into those failures. Didn't see those locally.

**Comment by srhb** [comment]
Weird.

```
❯ nix-prefetch-url https://github.com/kubernetes/kubernetes/commit/34ce573e9992ecdbc06dff1b4e3d0e9baa8353dd.patch
path is '/nix/store/51ji2x0akig7dc2v6llqmjk36xrfv99m-34ce573e9992ecdbc06dff1b4e3d0e9baa8353dd.patch'
189jajilys5zd0x4asbp6by8y3jl5hfaifqqjiq1h2bv00v1qdym
```

**Comment by srhb** [comment]
@grahamcofborg build kubernetes

**Comment by Mic92** [comment]
@GrahamcOfBorg test tests.kubernetes.dns.singlenode

File: nixos/release.nix [hunk 1]
```diff
@@ -292,7 +292,10 @@ in rec {
   tests.kernel-copperhead = callTest tests/kernel-copperhead.nix {};
   tests.kernel-latest = callTest tests/kernel-latest.nix {};
   tests.kernel-lts = callTest tests/kernel-lts.nix {};
-  tests.kubernetes = hydraJob (import tests/kubernetes/default.nix { system = "x86_64-linux"; });
+  tests.kubernetes.dns = callSubTestsOnTheseSystems ["x86_64-linux"] tests/kubernetes/dns.nix {};
+  ## kubernetes.e2e should eventually replace kubernetes.rbac when it works
+  #tests.kubernetes.e2e = callSubTestsOnTheseSystems ["x86_64-linux"] tests/kubernetes/e2e.nix {};
+  tests.kubernetes.rbac = callSubTestsOnTheseSystems ["x86_64-linux"] tests/kubernetes/rbac.nix {};
```
Comment (Mic92) [inline nixos/release.nix]: Let's remove some abstractions for our own sanity.

**Comment by Mic92** [comment]
@GrahamcOfBorg eval

File: nixos/release.nix [hunk 2]
```diff
@@ -300,6 +300,7 @@ in rec {
   ## kubernetes.e2e should eventually replace kubernetes.rbac when it works
   #tests.kubernetes.e2e = callSubTestsOnTheseSystems ["x86_64-linux"] tests/kubernetes/e2e.nix {};
   tests.kubernetes.rbac = callSubTestsOnTheseSystems ["x86_64-linux"] tests/kubernetes/rbac.nix {};
+  tests.kubernetes = callSubTestsOnTheseSystems ["x86_64-linux"] tests/kubernetes/default.nix {};
```
Comment (Mic92) [inline nixos/release.nix]: oh, wrong merge commit.

**Comment by srhb** [comment]
@Mic92 Great, thanks! Good to merge?

**Comment by Mic92** [comment]
I only fixed running tests from `release.nix`. I have not actually run all tests. If it works for you merging should be fine
@GrahamcOfBorg test kubernetes.dns kubernetes.rbac

**Comment by Mic92** [comment]
Also without tests it should be fine. Please also backport it to 18.03.

**Comment by Mic92** [comment]
@grahamc is the log output of tests above legit?

**Comment by grahamc** [comment]
This is so weird. Let's try again :o

@GrahamcOfBorg test kubernetes.dns kubernetes.rbac

**Comment by srhb** [comment]
@GrahamcOfBorg test kubernetes.dns.singlenode kubernetes.dns.multinode kubernetes.rbac.singlenode kubernetes.rbac.multinode

**Comment by srhb** [comment]
```
❯ nix-build nixos/release.nix -A tests.kubernetes.dns.singlenode -A tests.kubernetes.dns.multinode -A tests.kubernetes.rbac.singlenode -A tests.kubernetes.rbac.multinode
/nix/store/jnc32k3xixs12bchmpdwahcsn9hvb4k3-vm-test-run-kubernetes-dns-singlenode
/nix/store/rf5sw70z1g2hxq3zxh5k6jh4z7kagahd-vm-test-run-kubernetes-dns-multinode
/nix/store/yak4mp3j61mn74g89bj9blwkl8pcp2xb-vm-test-run-kubernetes-rbac-singlenode
/nix/store/jypgf9zbwhiwcpcqwlxlcpgd42c7dw35-vm-test-run-kubernetes-rbac-multinode
```

Looks good to me. @Mic92 Should I keep in the changes you made to release.nix or not? It seems they do nothing for ofborg, not sure about Hydra.

**Comment by grahamc** [comment]
These tests won't run on Hydra, just like they don't run on OfBorg.

I need to publish logs as to _why_ they don't evaluate obviously, but if you try them like I showed you'll see why:

```
[nix-shell:~/projects/nixpkgs]$ HOME=/homeless-shelter NIX_PATH=nixpkgs=$(pwd) nix-instantiate ./nixos/release.nix -A tests.kubernetes.dns.singlenode --option restrict-eval true --option build-timeout 1800 --arg supportedSystems  '["x86_64-linux"]' --show-trace
error: while evaluating the attribute 'singlenode' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:44:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/nixos/release.nix:30:25, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:52:
while evaluating 'hydraJob' at /home/grahamc/projects/nixpkgs/lib/customisation.nix:162:14, called from /home/grahamc/projects/nixpkgs/nixos/release.nix:30:28:
while evaluating the attribute 'outPath' at /home/grahamc/projects/nixpkgs/lib/customisation.nix:178:13:
while evaluating the attribute 'outPath' at /home/grahamc/projects/nixpkgs/lib/customisation.nix:147:13:
while evaluating the attribute 'buildCommand' of the derivation 'vm-test-run-kubernetes-dns-singlenode' at /home/grahamc/projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'buildCommand' of the derivation 'nixos-test-driver-kubernetes-dns-singlenode' at /home/grahamc/projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'buildCommand' of the derivation 'nixos-vm' at /home/grahamc/projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute 'system.build.toplevel' at /home/grahamc/projects/nixpkgs/nixos/modules/system/activation/top-level.nix:263:5:
while evaluating 'foldr' at /home/grahamc/projects/nixpkgs/lib/lists.nix:34:20, called from /home/grahamc/projects/nixpkgs/nixos/modules/system/activation/top-level.nix:134:12:
while evaluating 'fold'' at /home/grahamc/projects/nixpkgs/lib/lists.nix:37:15, called from /home/grahamc/projects/nixpkgs/lib/lists.nix:41:8:
while evaluating 'showWarnings' at /home/grahamc/projects/nixpkgs/nixos/modules/system/activation/top-level.nix:100:18, called from /home/grahamc/projects/nixpkgs/nixos/modules/system/activation/top-level.nix:107:16:
while evaluating 'foldr' at /home/grahamc/projects/nixpkgs/lib/lists.nix:34:20, called from /home/grahamc/projects/nixpkgs/nixos/modules/system/activation/top-level.nix:100:23:
while evaluating 'fold'' at /home/grahamc/projects/nixpkgs/lib/lists.nix:37:15, called from /home/grahamc/projects/nixpkgs/lib/lists.nix:41:8:
while evaluating the attribute 'warnings' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:44:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:75:45, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:52:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/modules.nix:312:9:
while evaluating the option `warnings':
while evaluating the attribute 'isDefined' at /home/grahamc/projects/nixpkgs/lib/modules.nix:344:5:
while evaluating 'filterOverrides' at /home/grahamc/projects/nixpkgs/lib/modules.nix:419:21, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:328:18:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:425:8:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:323:17:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:323:28, called from undefined position:
while evaluating 'dischargeProperties' at /home/grahamc/projects/nixpkgs/lib/modules.nix:386:25, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:324:62:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/modules.nix:216:48:
while evaluating the attribute 'config.warnings' at /home/grahamc/projects/nixpkgs/nixos/modules/system/boot/systemd.nix:702:5:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:224:10, called from undefined position:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/nixos/modules/system/boot/systemd.nix:702:51, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:224:16:
while evaluating 'optional' at /home/grahamc/projects/nixpkgs/lib/lists.nix:198:20, called from /home/grahamc/projects/nixpkgs/nixos/modules/system/boot/systemd.nix:703:7:
while evaluating the attribute 'serviceConfig.Type' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:44:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:75:45, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:52:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/modules.nix:312:9:
while evaluating the option `systemd.services.kube-apiserver.serviceConfig':
while evaluating the attribute 'mergedValue' at /home/grahamc/projects/nixpkgs/lib/modules.nix:339:5:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:339:32, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:339:19:
while evaluating 'merge' at /home/grahamc/projects/nixpkgs/lib/types.nix:267:20, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:342:8:
while evaluating 'mapAttrs' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:198:17, called from /home/grahamc/projects/nixpkgs/lib/types.nix:268:9:
while evaluating 'filterAttrs' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:115:23, called from /home/grahamc/projects/nixpkgs/lib/types.nix:268:35:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:116:18:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:116:29, called from undefined position:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/types.nix:268:51, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:116:62:
while evaluating the attribute 'ExecStart' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:334:7:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/types.nix:268:86, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:334:15:
while evaluating the attribute 'optionalValue' at /home/grahamc/projects/nixpkgs/lib/modules.nix:346:5:
while evaluating 'filterOverrides' at /home/grahamc/projects/nixpkgs/lib/modules.nix:419:21, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:328:18:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:425:8:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:323:17:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:323:28, called from undefined position:
while evaluating 'dischargeProperties' at /home/grahamc/projects/nixpkgs/lib/modules.nix:386:25, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:324:62:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/types.nix:273:55:
while evaluating the attribute 'ExecStart' at /home/grahamc/projects/nixpkgs/nixos/modules/services/cluster/kubernetes/default.nix:861:11:
while evaluating 'optionalString' at /home/grahamc/projects/nixpkgs/lib/strings.nix:138:26, called from /home/grahamc/projects/nixpkgs/nixos/modules/services/cluster/kubernetes/default.nix:863:15:
while evaluating the attribute 'etcd.caFile' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:44:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:75:45, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:52:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/modules.nix:312:9:
while evaluating the option `services.kubernetes.etcd.caFile':
while evaluating the attribute 'isDefined' at /home/grahamc/projects/nixpkgs/lib/modules.nix:344:5:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:333:19, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:333:14:
while evaluating the attribute 'value._type' at /home/grahamc/projects/nixpkgs/lib/modules.nix:424:73:
while evaluating the attribute 'value.content' at /home/grahamc/projects/nixpkgs/lib/modules.nix:482:14:
while evaluating the attribute 'default' at /home/grahamc/projects/nixpkgs/nixos/modules/services/cluster/kubernetes/default.nix:178:9:
while evaluating the attribute 'caFile' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:44:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:75:45, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:199:52:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/modules.nix:312:9:
while evaluating the option `services.kubernetes.caFile':
while evaluating the attribute 'isDefined' at /home/grahamc/projects/nixpkgs/lib/modules.nix:344:5:
while evaluating 'filterOverrides' at /home/grahamc/projects/nixpkgs/lib/modules.nix:419:21, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:328:18:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:425:8:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:323:17:
while evaluating anonymous function at /home/grahamc/projects/nixpkgs/lib/modules.nix:323:28, called from undefined position:
while evaluating 'dischargeProperties' at /home/grahamc/projects/nixpkgs/lib/modules.nix:386:25, called from /home/grahamc/projects/nixpkgs/lib/modules.nix:324:62:
while evaluating the attribute 'value' at /home/grahamc/projects/nixpkgs/lib/modules.nix:216:48:
while evaluating the attribute 'config.caFile' at /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/kubernetes-common.nix:11:5:
while evaluating the attribute 'passAsFile' of the derivation 'master-keys' at /home/grahamc/projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the derivation attribute 'name' at /home/grahamc/projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating 'noKey' at /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:26:11, called from /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:165:20:
while evaluating 'filterAttrs' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:115:23, called from /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:26:20:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:116:18:
while evaluating 'createSigningCertKey' at /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:52:26, called from /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:90:8:
while evaluating 'noCSR' at /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:25:11, called from /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:53:6:
while evaluating 'filterAttrs' at /home/grahamc/projects/nixpkgs/lib/attrsets.nix:115:23, called from /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:25:20:
while evaluating 'concatMap' at /home/grahamc/projects/nixpkgs/lib/lists.nix:102:18, called from /home/grahamc/projects/nixpkgs/lib/attrsets.nix:116:18:
while evaluating 'runWithCFSSL' at /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:9:24, called from /home/grahamc/projects/nixpkgs/nixos/tests/kubernetes/certs.nix:53:13:
access to path '/nix/store/xyi4nppbpshcfrgx4vlc827l3bz5qgfk-ca-cfss.json' is forbidden in restricted mode
```

In other words, the test seems to be using IFD which is not permitted.

**Comment by grahamc** [comment]
Specifically this is the problem, in certs.nix:

```
  runWithCFSSL = name: cmd:
    builtins.fromJSON (builtins.readFile (
      pkgs.runCommand "${name}-cfss.json" {
        buildInputs = [ pkgs.cfssl ];
      } "cfssl ${cmd} > $out"
    ));
```

the readFile is not permitted.

**Comment by grahamc** [comment]
I'm almost done fixing the test to be able to run on Hydra and OfBorg.

**Comment by grahamc** [comment]
The following patch seems to cause the tests to break, but is the right way to go to have this no longer violate IFD issues:

```diff
diff --git a/nixos/tests/kubernetes/certs.nix b/nixos/tests/kubernetes/certs.nix
index d3eff910c46..097488c46e0 100644
--- a/nixos/tests/kubernetes/certs.nix
+++ b/nixos/tests/kubernetes/certs.nix
@@ -7,11 +7,26 @@
 }:
 let
   runWithCFSSL = name: cmd:
-    builtins.fromJSON (builtins.readFile (
-      pkgs.runCommand "${name}-cfss.json" {
-        buildInputs = [ pkgs.cfssl ];
-      } "cfssl ${cmd} > $out"
-    ));
+    let secrets = pkgs.runCommand "${name}-cfss.json" {
+        buildInputs = [ pkgs.cfssl pkgs.jq ];
+        outputs = [ "out" "cert" "key" "csr" ];
+      }
+      ''
+        (
+          echo "${cmd}"
+          cfssl ${cmd} > tmp
+          cat tmp | jq -r .key > $key
+          cat tmp | jq -r .cert > $cert
+          cat tmp | jq -r .csr > $csr
+
+          touch $out
+        ) 2>&1 | fold -w 80 -s
+      '';
+    in {
+      key = secrets.key;
+      cert = secrets.cert;
+      csr = secrets.csr;
+    };
 
   writeCFSSL = content:
     pkgs.runCommand content.name {
@@ -25,10 +40,10 @@ let
   noCSR = content: pkgs.lib.filterAttrs (n: v: n != "csr") content;
   noKey = content: pkgs.lib.filterAttrs (n: v: n != "key") content;
 
-  writeFile = content: pkgs.writeText "content" (
-    if pkgs.lib.isAttrs content then builtins.toJSON content
-    else toString content
-  );
+  writeFile = content:
+    if pkgs.lib.isDerivation content
+    then content
+    else pkgs.writeText "content" (builtins.toJSON content);
 
   createServingCertKey = { ca, cn, hosts? [], size ? 2048, name ? cn }:
     noCSR (

```

**Comment by Mic92** [comment]
@srhb can you add this?

**Comment by srhb** [comment]
Honestly, this PR is a complete mess now. I'd much rather focus on getting the patch that unbreaks kube-proxy merged and have a separate PR or issue regarding tests being broken on Hydra and ofborg. Thanks for all your input.

**Comment by srhb** [comment]
OK, all the test changes are now in #37199 and removed from this PR.

This PR should be good to merge in order to fix kubernetes networking.

I doubt I'll be able to finish making the test non-IFD-compatible. I've spent a few hours trying to track down the (last?) stray issue(s) without luck.

**Comment by cstrahan** [comment]
Applied in e8cbd44 -- thanks!

**Comment by itorres** [comment]
Thanks a lot for the patch.

Is there an ETA for this patch to get it backported to 18.03? I am hitting this issue in a new  install.

**Comment by azazel75** [comment]
Hey @itorres, 
I had the same problem and I resolved using this trick from the wiki https://nixos.wiki/wiki/FAQ#How_can_I_install_a_package_from_unstable_while_remaining_on_the_stable_channel.3F
just yesterday
I hope it helps

**Pull request closed without merging**